File: MergeList.java
// Time Complexity : O(N * log k), k = number of lists, N = total number of nodes in all list (n * k), n = avg length of each list
// Space Complexity :O(k), height of the tree which will be k elements in queue.
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english : Maintain a min heap of size K to find the miniumum element while traversing.
// We did a heap sort

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a,b) -> (a.val - b.val));
        ListNode dummy = new ListNode(-1);
        for(ListNode head : lists){
            if(head != null){
                pq.add(head);
            }
        }

        ListNode curr = dummy;
        while(!pq.isEmpty()){
            ListNode el = pq.poll();
            curr.next = el;
            curr = curr.next;

            if(el.next != null){
                pq.add(el.next);
            }    
        }
        
        return dummy.next;
    }
}


// Another Approach
// Time Complexity : O(N * k) or O(nk^2), k = number of lists, N = total number of nodes in all list (n * k), n = avg length of each list
// Space Complexity :O(1), no extra auxiallry space being used
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english : Merge every list one by one

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode result = new ListNode(Integer.MIN_VALUE);
        
        for(ListNode head : lists){
            if(head != null){
                result = merge(result,head);
            }     
        }
        return result.next;
    }

    private ListNode merge(ListNode h1 , ListNode h2){
        ListNode dummy = new ListNode(-1);
        ListNode curr = dummy;

        while(h1 != null && h2 != null){
            if(h1.val < h2.val){
                curr.next = h1;
                h1 = h1.next;
            }
            else{
                curr.next = h2;
                h2 = h2.next; 
            }

            curr = curr.next;
        }

        if(h1 != null){
            curr.next = h1;
        }

        if(h2 != null){
            curr.next = h2;
        }


        return dummy.next;
    }
}

File: mergelist.py
# // Time Complexity :O(nklog(n))
# // Space Complexity :O(k),elements in the heap
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :no


# // Your code here along with comments explaining your approach
from heapq import heappush as insert
from heapq import heappop as remove
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        maxheap=[]
        for i in range(len(nums)):
            insert(maxheap,(-nums[i]))
            # print(maxheap)
            if (len(maxheap)-1)>(len(nums)-k):
                remove(maxheap)
        # print(maxheap)
            
        return -(maxheap[0])
        




from heapq import heappush as insert
from heapq import heappop as remove
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class CustomListNode:
    def __init__(self, node):
        self.node = node
        # self.next = next
    
    def __lt__(self, other):
            return self.node.val < other.node.val
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) ->Optional[ListNode]:
        dummy=ListNode(-1)
        curr=dummy
        pq=[]
        
        
        for i in lists:
            if i is not None:
                print(i.val)
            
                insert(pq,(i.val,CustomListNode(i)))
        while pq:
            
            minim=remove(pq)
            print(minim[1].node)
            nnode=minim[1].node
        
            curr.next=nnode
            curr=curr.next
            if nnode.next is not None:
                
                insert(pq,(nnode.next.val,CustomListNode(nnode.next)))
        return dummy.next

        
        
        